### DISCUSSION TITLE
How to use Packer to create just-in-time self-hosted runners?

### DISCUSSION THREAD

**Discussion by bill-taut** [community/community#82388] [Actions] [answered]
### Select Topic Area

Question

### Body

I am trying to create a VM image in Packer that can be used as a **just-in-time** self-hosted runner.

I am able already able to create a normal (not just-in-time) self-hosted runner just fine by having the following in my setup.sh:

```
cd /opt
sudo mkdir github
sudo chown runner github
cd github
mkdir actions-runner && cd actions-runner
curl -o actions-runner-linux-x64-2.311.0.tar.gz -L https://github.com/actions/runner/releases/download/v2.311.0/actions-runner-linux-x64-2.311.0.tar.gz
tar xzf ./actions-runner-linux-x64-2.311.0.tar.gz
./config.sh --url https://github.com/taut-ai/taut --token <self-hosted runner token from Github UI>

cd /etc
sudo echo "#!/bin/sh -e" > ~/rc.local
sudo echo "su runner -c 'nohup /opt/github/actions-runner/run.sh >/dev/null 2>&1 &'" >> ~/rc.local
sudo echo "exit 0" >> ~/rc.local
sudo mv ~/rc.local .
sudo chmod a+x rc.local
```

This has the obvious drawback, though, that the image can really only be used on one self-host runner, since the runner token is pre-baked into the image.

This seems like a natural use case for a just-in-time self-hosted runner, but I cannot figure out how to do this through Packer. Does anyone have suggestions for how to do this?

**Comment by crohr** [comment, marked as answer]
You need to inject the runner token a boot time, not at build time. Also if you're on AWS look at [RunsOn](https://runs-on.com) for an easy solution.

**Comment by bill-taut** [comment]
Agreed, thanks. I've learned a lot about this since asking this question.
